gpkit/tools.py

```diff
@@ -112,3 +115,53 @@ def mdmake(filename, make_tex=True):
         with open(filename+".tex.md", "w") as f:
             f.write(texmd)
     return open(filename+".py")
+
+
+def bound_all_variables(model, eps=1e-30, lower=None, upper=None):
+    "Returns model with additional constraints bounding all free variables"
+    lb = lower if lower else eps
+    ub = upper if upper else 1/eps
+    constraints = []
+    for varkey in model.varkeys:
+        units = varkey.descr.get("units", 1)
+        constraints.append([ub*units >= Variable(**varkey.descr),
+                            Variable(**varkey.descr) >= lb*units])
+    m = Model(model.cost, [constraints, model], model.substitutions)
+    m.bound_all = {"lb": lb, "ub": ub, "varkeys": model.varkeys}
+    return m
+
+
+# pylint: disable=too-many-locals
+def determine_unbounded_variables(model, solver=None, verbosity=0,
+                                  eps=1e-30, lower=None, upper=None, **kwargs):
+    "Returns labeled dictionary of unbounded variables."
+    m = bound_all_variables(model, eps, lower, upper)
+    sol = m.solve(solver, verbosity, **kwargs)
+    lam = sol["sensitivities"]["la"][1:]
+    out = defaultdict(list)
+    for i, varkey in enumerate(m.bound_all["varkeys"]):
+        lam_gt, lam_lt = lam[2*i], lam[2*i+1]
```

Comment:
bound_all["varkeys"] is declared above, I think?
On Mar 25, 2016 12:42, "Warren Hoburg" <notifications@github.com> wrote:

> In gpkit/tools.py
> <https://github.com/hoburg/gpkit/pull/571#discussion_r57461872>:
>
> > +                            Variable(**varkey.descr) >= lb*units])
> > +    m = Model(model.cost, [constraints, model], model.substitutions)
> > +    m.bound_all = {"lb": lb, "ub": ub, "varkeys": model.varkeys}
> > +    return m
> > +
> > +
> > +# pylint: disable=too-many-locals
> > +def determine_unbounded_variables(model, solver=None, verbosity=0,
> > +                                  eps=1e-30, lower=None, upper=None, **kwargs):
> > +    "Returns labeled dictionary of unbounded variables."
> > +    m = bound_all_variables(model, eps, lower, upper)
> > +    sol = m.solve(solver, verbosity, **kwargs)
> > +    lam = sol["sensitivities"]["la"][1:]
> > +    out = defaultdict(list)
> > +    for i, varkey in enumerate(m.bound_all["varkeys"]):
> > +        lam_gt, lam_lt = lam[2*i], lam[2*i+1]
>
> really?
>
> In [16]: from gpkit import Variable, Model
>
> In [17]: x = Variable('x')
>
> In [18]: m = Model(x, [x >= 1])
>
> In [19]: m.varkeys
> Out[19]: {x: {x}}
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/hoburg/gpkit/pull/571/files/df5a416930d67c14369ba442ac71ae614255fc53#r57461872>
>